File: src/chapters/differential-equations/content/code/taylor3.rs
// Third-order Taylor method (first two total derivatives of f).
fn taylor3<F, G, H>(f: F, df: G, d2f: H, t0: f64, y0: f64, h: f64, n: usize) -> f64
where F: Fn(f64, f64) -> f64, G: Fn(f64, f64) -> f64, H: Fn(f64, f64) -> f64 {
    let (mut t, mut y) = (t0, y0);
    for _ in 0..n { y += h * f(t, y) + h * h / 2.0 * df(t, y) + h * h * h / 6.0 * d2f(t, y); t += h; }
    y
}
fn main() {
    let f = |t: f64, y: f64| 2.0 * y - 10.0 * t * t + 2.0 * t;
    let df = |t: f64, y: f64| 4.0 * y - 20.0 * t * t - 16.0 * t + 2.0;
    let d2f = |t: f64, y: f64| 8.0 * y - 40.0 * t * t - 32.0 * t - 16.0;
    println!("y(1) = {}", taylor3(f, df, d2f, 0.0, 1.0, 0.1, 10));
}

File: src/pages/practice/guides/matlab/taylor3.m
function y = taylor3(f, df, d2f, t0, y0, h, n)
% TAYLOR3  Third-order Taylor method (first two total derivatives of f).
    t = t0; y = y0;
    for i = 1:n
        y = y + h*f(t, y) + h^2/2*df(t, y) + h^3/6*d2f(t, y);
        t = t + h;
    end
end

% --- Demo ---
f   = @(t, y) 2*y - 10*t^2 + 2*t;
df  = @(t, y) 4*y - 20*t^2 - 16*t + 2;
d2f = @(t, y) 8*y - 40*t^2 - 32*t - 16;
fprintf('y(1) = %.6f\n', taylor3(f, df, d2f, 0, 1, 0.1, 10));
